### ISSUE TITLE
eigen 3.4.0 upgrade fails: SHA256 mismatch

### ISSUE THREAD

**Issue by maspotts** [Homebrew/homebrew-core#84881]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
The eigen.rb formula downloads this package:

  https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.0.tar.gz

whose SHA256 is:

 Actual: a5f198912559739eaa5a0857978edcbdf733832c39d733ee5cf83193d314a2ac

but the formula expects this SHA256:

  Expected: 575d079f9fb56806f605b90fd85fbaeb4a3fb31b573fbbe1ab2fed3b5f88d8fe

hence mismatch.  I verified the actual SHA256 by downloading directly.
```

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

brew upgrade eigen

### What happened (include all command output)?

$ brew upgrade eigen
Updating Homebrew...
==> Upgrading 1 outdated package:
eigen 3.3.9 -> 3.4.0
==> Downloading https://ghcr.io/v2/homebrew/core/cmake/manifests/3.21.2
Already downloaded: /Users/mike/Library/Caches/Homebrew/downloads/e53caefe5de8095f304f2da999d9a73844e1359e599235adc9ab82512ffe4b1a--cmake-3.21.2.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/cmake/blobs/sha256:a039c944e143dd97b05b87759d78cc491bc294336798ff4e141feec878303200
Already downloaded: /Users/mike/Library/Caches/Homebrew/downloads/e676ef38a64e298cc15d49e2786658f96192daa0cf41353b4d3646a78ee903ad--cmake--3.21.2.catalina.bottle.tar.gz
Warning: Building eigen from source as the bottle needs:
- HOMEBREW_CELLAR: /usr/local/Cellar (yours is /opt/brew/Cellar)
- HOMEBREW_PREFIX: /usr/local (yours is /opt/brew)
==> Downloading https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.0.tar.gz
Already downloaded: /Users/mike/Library/Caches/Homebrew/downloads/f3defcd3bc840e23553cd522a7b0f554bbce3aa157fa09ab0cbe14da5aaed716--eigen-3.4-3395f4e604bde4006ff963aeeab271f11f3c2af7.tar.gz
Error: SHA256 mismatch
Expected: 575d079f9fb56806f605b90fd85fbaeb4a3fb31b573fbbe1ab2fed3b5f88d8fe
 Actual: a5f198912559739eaa5a0857978edcbdf733832c39d733ee5cf83193d314a2ac
   File: /Users/mike/Library/Caches/Homebrew/downloads/f3defcd3bc840e23553cd522a7b0f554bbce3aa157fa09ab0cbe14da5aaed716--eigen-3.4-3395f4e604bde4006ff963aeeab271f11f3c2af7.tar.gz
To retry an incomplete download, remove the file above.

### What did you expect to happen?

Successful upgrade.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew upgrade eigen
```

**Comment by carlocab** [comment]
Please don’t omit information requested in the issue template. What is the output of `brew config` and `brew doctor`?

**Comment by maspotts** [comment]
Sorry: didn't pick up on those instructions first time around!

```
$ brew config
HOMEBREW_VERSION: 3.2.11
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 84f4fbf3cf5e83d33de0b28b2658fcb50aaa761e
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b88b6b36cb475c58956ed0825252a43a4d809359
Core tap last commit: 57 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/brew
HOMEBREW_REPOSITORY: /opt/brew
HOMEBREW_CELLAR: /opt/brew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vi
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_ANALYTICS: set
Homebrew Ruby: 2.6.3 => /opt/brew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit broadwell
Clang: 12.0.0 build 1200
Git: 2.31.1 => /opt/brew/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: 12.4
```



```
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ilmbase

Warning: Your Homebrew's prefix is not /usr/local.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/usr/local).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.
```

I don't know how to fix the ilmbase issue since it's a dependency of multiple install formulae.  I can't fix the /usr/local (prefix) issue because I don't want to mix homebrew with other installed libraries in /usr/local .  But I don't believe my prefix shoudl be causing this issue.

**Comment by maspotts** [comment]
Verifying actual sha256 hash:

```
$ curl "https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.0.tar.gz" | sha256sum
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 2647k    0 2647k    0     0   705k      0 --:--:--  0:00:03 --:--:--  705k
a5f198912559739eaa5a0857978edcbdf733832c39d733ee5cf83193d314a2ac  -

```

**Comment by gspencergoog** [comment]
I have the same issue, and can verify that my SHA256 is the same when downloaded via curl, but I get a different "Actual" SHA256 than the OP. I can't run `brew gist-logs` because it just complains at me about the command line tools. 

```
% brew upgrade eigen
==> Upgrading 1 outdated package:
eigen 3.3.7 -> 3.4.0
Warning: Building eigen from source as the bottle needs:
- HOMEBREW_CELLAR: /usr/local/Cellar (yours is /Users/gspencer/homebrew/Cellar)
- HOMEBREW_PREFIX: /usr/local (yours is /Users/gspencer/homebrew)
==> Downloading https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.0.tar.
Already downloaded: /Users/gspencer/Library/Caches/Homebrew/downloads/f3defcd3bc840e23553cd522a7b0f554bbce3aa157fa09ab0cbe14da5aaed716--eigen-3.4-3395f4e604bde4006ff963aeeab271f11f3c2af7.tar.gz
Error: SHA256 mismatch
Expected: 575d079f9fb56806f605b90fd85fbaeb4a3fb31b573fbbe1ab2fed3b5f88d8fe
  Actual: 0a5d7ebd1f0b50ae54e73e5acea296ba221e211b5e0df6143c3af8e1e3eee87e
    File: /Users/gspencer/Library/Caches/Homebrew/downloads/f3defcd3bc840e23553cd522a7b0f554bbce3aa157fa09ab0cbe14da5aaed716--eigen-3.4-3395f4e604bde4006ff963aeeab271f11f3c2af7.tar.gz
```

`brew config` output:
```
HOMEBREW_VERSION: 3.2.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: 84f4fbf3cf5e83d33de0b28b2658fcb50aaa761e
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core.git
Core tap HEAD: 9b586610d449e50571c188bfdb31fab9466eec75
Core tap last commit: 56 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /Users/gspencer/homebrew
HOMEBREW_REPOSITORY: /Users/gspencer/homebrew
HOMEBREW_CELLAR: /Users/gspencer/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.8dRHKOhLOS/org.macosforge.xquartz:0
HOMEBREW_EDITOR: my_open_mate_wait
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /Users/gspencer/homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: octa-core 64-bit skylake
Clang: 11.0.0 build 1100
Git: 2.27.0 => /Users/gspencer/homebrew/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.14.6-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: 11.3.1
XQuartz: 2.7.11 => /opt/X11
```

`brew doctor` output:
```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/more/.
You should download the Command Line Tools for Xcode 11.3.1.

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ilmbase

Warning: Your Homebrew's prefix is not /usr/local.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/usr/local).
You will encounter build failures with some formulae.
```

I have tried removing the downloaded image and re-downloading it multiple times.
I have already updated command line tools to the latest available for my OS version (Mojave: 10.14.6), but brew keeps reporting it as out of date.  I'll try updating my OS version and XCode if you think that's the cause (unlikely).
I can't upgrade `ilmbase` because I run into this issue, since `eigen` is a dependency.

**Comment by carlocab** [comment]
We should check why the checksum changed, because that was verified at the time the formula was updated. If the change was intentional, then we can update the formula.

That said, I think you shouldn't even be seeing this error, because `eigen`'s bottle is `any_skip_relocation`:

https://github.com/Homebrew/homebrew-core/blob/c922d9d577f4185da948a0961e43690ce35fa05d/Formula/eigen.rb#L14-L16

This means that there should be no need to build this from source. I suspect `brew` is being tripped up by the `all` bottle here. Does that sound right, @MikeMcQuaid?

**Comment by MikeMcQuaid** [comment]
> This means that there should be no need to build this from source. I suspect `brew` is being tripped up by the `all` bottle here. Does that sound right, @MikeMcQuaid?

Yeh, it sounds like this should be built from source, agreed. No idea why the `all` bottle isn't working as expected but sounds like a 🐛 if so.

**Comment by maspotts** [comment]
Any idea what the resolution should be here?  Should I just make a local copy of the eigen formula and use it instead, or is someone likely to update it shortly?  Does someone need to get in touch with the eigen maintainers and verify the new MD5?  (Can I help with that?)

**Comment by carlocab** [comment]
> Does someone need to get in touch with the eigen maintainers and verify the new SHA256?  (Can I help with that?)

Please do help. When upstream verifies the new checksum we can update the formula.

**Comment by maspotts** [comment]
I talked to another homebrew formula maintainer (chenrui333) who asked me to follow up with the eigen team: I connected with their official discord channel and discovered that the current formula is using the wrong url: it should be:

  https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.tar.gz

whose SHA256 is:

  8586084f71f9bde545ee7fa6d00288b264a2b7ac3607b974e54d13e7162c1c72

For provenance please see their official releases page:

  https://gitlab.com/libeigen/eigen/-/releases#3.4.0

Is this sufficient evidence to update the formula and close this bug?

**Comment by carlocab** [comment]
I'm confused. Isn't this the current URL in the formula?

https://github.com/Homebrew/homebrew-core/blob/1ce3271cff8ae34b6c297c11931e92a47b9130d7/Formula/eigen.rb#L4

This matches the correct URL exactly. What is the other URL that you are referring to?

**Comment by maspotts** [comment]
Not quite: this is the url in the formula, as you say:

` url "https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.0.tar.gz" `

but this is what it should be:

` url "https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.tar.gz" `

The correct (new) url includes "/3.4.0/" whereas the current url includes "/3.4/".  The two urls return different packages.  The eigen maintainers are adamant that the "/3.4.0/" variant is correct (and that's what they list in the release page); they couldn't figure out what the "/3.4/" variant resolves to but they are adamant that it's not the correct release package.

**Comment by carlocab** [comment]
Huh, my mistake. Yes, I think that should be fine. Do you mind opening a pull request to update the formula? Please also bump the formula's `revision`, as it seems we also need to update the users who currently have this installed. Just add `revision 1` to the line right after `license`.

**Comment by jonchang** [comment]
3.4 resolves to a branch name. Note that github doesn't have this issue because tags and branches have a different url structure e.g. https://github.com/actions/runner/archive/refs/tags/v2.282.0.tar.gz (vs refs/heads for branches)

**Comment by maspotts** [comment]
I've updated my copy of https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/eigen.rb .  How do I test it locally before committing and generating the PR?

**Comment by carlocab** [comment]
You can do
```
brew install -s eigen
brew test eigen
brew audit --strict eigen
```

**Comment by maspotts** [comment]
Thanks: but also: how do I tell brew to install using my local (fixed) formula?  AFAIK "brew install -s eigen" will use the formula from github, not my local copy.  Can I use (eg) "brew install -s /path/to/my/eigen.rb" ?

**Comment by carlocab** [comment]
Just do
```
cp /path/to/my/eigen.rb $(brew --repo homebrew/core)/Formula/eigen.rb
```
before doing the commands I mentioned above.

**Comment by maspotts** [comment]
Success:
```
brew install -s eigen
eigen 3.3.9 is already installed but outdated
==> Downloading https://ghcr.io/v2/homebrew/core/cmake/manifests/3.21.2
Already downloaded: /Users/mike/Library/Caches/Homebrew/downloads/e53caefe5de8095f304f2da999d9a73844e1359e599235adc9ab82512ffe4b1a--cmake-3.21.2.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/cmake/blobs/sha256:a039c944e143dd97b05b87759d78cc491bc294336798ff4e141feec878303200
Already downloaded: /Users/mike/Library/Caches/Homebrew/downloads/e676ef38a64e298cc15d49e2786658f96192daa0cf41353b4d3646a78ee903ad--cmake--3.21.2.catalina.bottle.tar.gz
==> Downloading https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.tar.gz
  # #=O=#   #
==> Upgrading eigen
  3.3.9 -> 3.4.0

==> Installing dependencies for eigen: cmake
==> Installing eigen dependency: cmake
==> Pouring cmake--3.21.2.catalina.bottle.tar.gz
🍺  /opt/brew/Cellar/cmake/3.21.2: 6,615 files, 69.2MB
==> Installing eigen
==> cmake -Dpkg_config_libdir=/opt/brew/Cellar/eigen/3.4.0/lib ..
==> make install
🍺  /opt/brew/Cellar/eigen/3.4.0: 546 files, 8.4MB, built in 33 seconds
Removing: /opt/brew/Cellar/eigen/3.3.9... (493 files, 6.6MB)
```

**Comment by maspotts** [comment]
For some reason brew is continuing to upgrade all outdated packages, so I'll run the other two commands once it's done...

**Comment by maspotts** [comment]
```
brew test eigen
==> Testing eigen
==> /usr/bin/clang++ test.cpp -I/opt/brew/Cellar/eigen/3.4.0/include/eigen3 -o test
==> ./test
```

```
brew audit --strict eigen
eigen:
  * Formula eigen contains non-standard SPDX licenses: ["MPL-2.0 revision 1"].
    For a list of valid licenses check: https://spdx.org/licenses/
Error: 1 problem in 1 formula detected
```

Did I mess up the license string?

**Comment by maspotts** [comment]
OK I figured out my mistake: I misinterpreted the "revsion 1" instructions. I've fixed that and now all looks good:

**Comment by maspotts** [comment]
I'm having trouble pushing my branch using my personal access token.  I tried to provide it:

```
$ git remote set-url origin https://<mytoken>@github.com/Homebrew/homebrew-core.git
```

but the push failed:

```
$ git push --set-upstream origin fix-eigen-3.4.0-url
remote: Permission to Homebrew/homebrew-core.git denied to maspotts.
fatal: unable to access 'https://github.com/Homebrew/homebrew-core.git/': The requested URL returned error: 403
```

I know this is off-topic but I can't proceed unless I can get past it.  Alternatively I'm happy to paste the new formula into this ticket and one of you can take of the merge (although I'd prefer to figure this problem out in case I hit it again in the future!).

**Comment by Bo98** [comment]
You need to create a fork and then push to that fork.

**Comment by maspotts** [comment]
So I keep my checked-out copy of the primary repo, and then add my fork as a new remote?  Then push my new branch to my fork?  Then the PR will be to merge from <myfork>/<mybranch> to <primary>/master ?  (is that possible?)

**Comment by Bo98** [comment]
That should work, yes. It's a common scenario, for example:

![](https://i.imgur.com/eFT2Rsn.png)

**Comment by maspotts** [comment]
cool: thanks!  will work on it later today...

**Comment by maspotts** [comment]
Created PR: https://github.com/Homebrew/homebrew-core/pull/85256

**Comment by carlocab** [comment]
Closed in 7d475fc3e7135bd394fc9e163936f983436c839b. Thanks @maspotts!
